services/headless-lms/chatbot/src/chatbot_tools/course_progress.rs

```diff
@@ -0,0 +1,650 @@
+use headless_lms_models::{
+    course_modules::{CompletionPolicy, CourseModule},
+    user_exercise_states::UserCourseProgress,
+};
+use headless_lms_utils::prelude::BackendError;
+use sqlx::PgConnection;
+
+use crate::{
+    azure_chatbot::ChatbotUserContext,
+    chatbot_tools::{AzureLLMToolDefinition, ChatbotTool, LLMTool, LLMToolType, ToolProperties},
+    prelude::{ChatbotError, ChatbotErrorType, ChatbotResult},
+};
+
+pub type CourseProgressTool = ToolProperties<CourseProgressState, CourseProgressArguments>;
+
+impl ChatbotTool for CourseProgressTool {
+    type State = CourseProgressState;
+    type Arguments = CourseProgressArguments;
+
+    fn parse_arguments(_args_string: String) -> ChatbotResult<Self::Arguments> {
+        Ok(CourseProgressArguments {})
+    }
+
+    /// Create a CourseProgressTool instance
+    async fn from_db_and_arguments(
+        conn: &mut PgConnection,
+        arguments: Self::Arguments,
+        user_context: &ChatbotUserContext,
+    ) -> ChatbotResult<Self> {
+        let user_progress = headless_lms_models::user_exercise_states::get_user_course_progress(
+            conn,
+            user_context.course_id,
+            user_context.user_id,
+        )
+        .await?;
+        let modules =
```

Comment:
We should check that we don't tell the progress for additional modules that are not open yet. (All chapters in module have opens_at set to the future.) This should also be unit tested.